lib/Dialect/Substrait/Transforms/EmitDeduplication.cpp

```diff
@@ -257,6 +259,13 @@ struct PushDuplicatesThroughFilterPattern : public OpRewritePattern<FilterOp> {
     deduplicateRegionArgs(newOp.getCondition(), emitOp.getMapping(),
                           newInput.getType(), rewriter);
 
+    // Deduplicating block args may create common subexpressions. Eliminate
+    // them immediately.
+    {
+      DominanceInfo domInfo;
+      mlir::eliminateCommonSubExpressions(rewriter, domInfo, newOp);
```

Comment:
This seems like it could be expensive to do inside a pattern. But not blocking, we can re-evaluate later.

Reply:
I have thought a bit about this: I think that this is the place with the lowest algorithmic complexity. The reason is the following: during duplicate elimination of the ops that produce the inputs of this op, some arguments into this region may be deduplicated, i.e., get fused. This, in turn, may create common sub-expressions since `field_references` that were previously referring to two different fields now refer to the same (deduplicated) field. And that, in turn, may allow to deduplicate the *outputs* of this op by eliminating these common sub-expressions using CSE, which are then interesting for the next op.

If we don't run CSE *inside* the op, we could have to run the CSE and emit deduplication passes as many times as the use-def tree is high because each pass could unlock something that the other pass could clean up. That way, we would have to go over the whole IR *every time* we repeat that cycle. With the current approach, emit deduplication goes over it once and CSE goes over the inner regions (only!) once every time inputs of that op got deduplicated (and only then!).

Let me document that somewhere in the code...

Reply:
(Updated git commit message.)